cordova-lib/src/cordova/platform.js

```diff
@@ -78,62 +76,119 @@ function add(hooksRunner, projectRoot, targets, opts) {
     }
 
     return hooksRunner.fire('before_platform_add', opts)
-    .then(function() {
-        return promiseutil.Q_chainmap(targets, function(t) {
-            // For each platform, download it and call its "create" script.
-
-            var p;  // The promise to be returned by this function.
-            var platform = t.split('@')[0];
-            // If t is not a platform or platform@version, it must be a dir.
-            // In this case get platform name from package.json in that dir and
-            // skip lazy-load.
-            if( !(platform in platforms) ) {
-                var pPath = path.resolve(t);
-                var pkg;
-                // Prep the message in advance, we might need it in several places.
-                msg = 'The provided path does not seem to contain a ' +
-                      'Cordova platform: ' + t;
-                try {
-                    pkg = require(path.join(pPath, 'package'));
-                } catch(e) {
-                    throw new CordovaError(msg + '\n' + e.message);
-                }
-                if ( !pkg || !pkg.name ) {
-                    throw new CordovaError(msg);
-                }
-                // Package names for Cordova platforms look like "cordova-ios".
-                var nameParts = pkg.name.split('-');
-                var name = nameParts[1];
-                if (name == 'amazon') {
-                    name = 'amazon-fireos';
-                }
-                if( !platforms[name] ) {
-                    throw new CordovaError(msg);
-                }
-                platform = name;
+	.then(function () {
+	    return promiseutil.Q_chainmap(targets, function (target) {
+
+	        // For each platform, download it and call its "create" script.
+	        var parts = target.split('@');
+	        var platform = parts[0];
+	        var version = parts[1];
+
+	        return Q.when().then(function () {
+	            if (!(platform in platforms)) {
+	                return getPlatformDetailsFromDir(target);
+	            }
+	            else {
+	                if (!version) {
+	                    events.emit('verbose', 'No version supplied. Retrieving version from config.xml...');
+	                }
+	                version = version || getVersionFromConfigFile(platform, cfg);
+	                if (version) {
+	                    version = version.replace(/^file:\/\//, ''); // If version is a folder, remove the 'file://' prefix from it.
```

Comment:
Couldn't figure out how to reply to old comment, so creating a new one again :S.

If it starts with "file:", then it's a URL (with file: being the scheme). If it has any `:`s, or spaces, then it's a malformed URL that most parses allow. However, we really should support properly formatted URLs by using the url.parse method.